[SUREFIRE-1964] Support for method filtering on excludesFile and includesFile

imonteroperez opened the pull request:
### Highlights
- See [SUREFIRE-1964](https://issues.apache.org/jira/browse/SUREFIRE-1964)
- This PR provides support for method filtering on `excludesFile` and `includesFile`. 
- If we perform `mvn test -Dsurefire.excludesFile=/foo/bar/file` and it is using method filtering like:
```
test.SomeTest#test1
test.SomeTest#test2
test.SomeTest#test3
test.SomeTest#test4
test.SomeTest#test5
test.SomeTest#test6
```
- Before this PR then we obtain `Method filter prohibited in includes|excludes|includesFile|excludesFile parameter`
- After this PR method filter is applied, so we obtain:
```
[INFO] --- maven-surefire-plugin:3.0.0-M6-SNAPSHOT:test (default-test) @ simple-maven-project-with-tests ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running test.SomeTest
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.035 s - in test.SomeTest
[INFO] Running test.OtherTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in test.OtherTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.808 s
[INFO] Finished at: 2021-12-03T18:00:37+01:00
[INFO] ------------------------------------------------------------------------
```


Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SUREFIRE) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SUREFIRE-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `SUREFIRE-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean install`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

slawekjaranowski commented:
@imonteroperez  please rebase and resolve conflicts

imonteroperez commented:
> @imonteroperez please rebase and resolve conflicts

Updated on https://github.com/apache/maven-surefire/pull/400/commits/abfd73f25e29798a47454c5dc643ff2967e6bc86 and no conflicts obtained

slawekjaranowski commented:
Please also squash to one finish commit
conflicts still exist ... `This branch cannot be rebased due to conflicts`

olamy commented:
@slawekjaranowski I cannot see any conflicts and btw we can squash when merging..

slawekjaranowski reviewed (changes requested):
Documentation for new feature should be also updated ... 

eg. on page https://maven.apache.org/surefire/maven-surefire-plugin/examples/inclusion-exclusion.html  or create another one page.

on plugin params `includesFile`, `excludesFile` put link to page.

In other case no one will know about it.

slawekjaranowski commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java:
```diff
@@ -2282,32 +2296,113 @@ private void maybeAppendList( List<String> base, List<String> list )
+            {
+                sb.append( "," ).append( include );
+            }
+            IncludeExcludeList excludeList = getExcludeList();
+            List<String> excludedTestMethods = excludeList.getTestMethods();
+            if ( !excludedTestMethods.isEmpty() ) 
+            {
+                includeList.clear();
+                includeList.addAll( excludedTestMethods );
+                for ( String method : excludedTestMethods ) 
+                {
+                  sb.append( "," ).append( method );
+                }
+            }
+            if ( !Arrays.equals( 
+                    Collections.singletonList( getDefaultExcludes() )
+                        .toArray( new String[0] ),
+                    excludeList.getTestClasses().toArray( new String[0] ) ) )
+            {
+                for ( String clazz : excludeList.getTestClasses() ) 
+                {
+                    sb.append( "," ).append( clazz );
+                }
+            }
+            
+            if ( sb.length() == 0 ) 
+            {
+                includedExcludedTests = new TestListResolver( 
+                        asList ( getDefaultIncludes() ), 
+                        Collections.singletonList( getDefaultExcludes() ) );
+            } 
+            else 
+            {
+                String testToIncludeExclude = sb.deleteCharAt( 0 ).toString( );
+                if ( !Arrays.equals( getDefaultIncludes(), split( testToIncludeExclude, "," ) ) ) 
+                {
+                    setTest( testToIncludeExclude );
+                }
+                includedExcludedTests = new TestListResolver( includeList, excludeList.getTestClasses() );
+            }
```
There are many decision branches here.

Maybe it will be possible put this logic in `TestListResolver` class and prepare unit test for it to cover all branches.

Such tests can show which input params impact to result list.

slawekjaranowski commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java:
```diff
@@ -2271,7 +2324,8 @@ private void maybeAppendList( List<String> base, List<String> list )
 
-            checkMethodFilterInIncludesExcludes( includes );
+            List<String> methodFilterInIncludes = getMethodFilterInIncludesExcludes( includes );
+            addAll( includes, methodFilterInIncludes.toArray( new String[0] ) );
```
`includes` can be reference to anything - should be immutable

slawekjaranowski commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java:
```diff
@@ -2282,32 +2336,113 @@ private void maybeAppendList( List<String> base, List<String> list )
+            List<String> excludedTestMethods = excludeList.getTestMethods();
+            if ( !excludedTestMethods.isEmpty() ) 
+            {
+                includeList.clear();
```
`includeList` is return value from `getIncludeList` should be immutable ...

slawekjaranowski commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java:
```diff
@@ -2282,32 +2336,113 @@ private void maybeAppendList( List<String> base, List<String> list )
+
+        IncludeExcludeList( List<String> testClasses, List<String> testMethods ) 
+        {
+            super();
```
needed?

slawekjaranowski commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java:
```diff
@@ -1839,7 +1841,7 @@ private PluginFailureReason getEffectiveFailIfNoTests()
         if ( isSpecificTestSpecified() )
         {
-            return getFailIfNoSpecifiedTests() ? COULD_NOT_RUN_SPECIFIED_TESTS : NONE;
+            return getFailIfNoSpecifiedTests() ? COULD_NOT_RUN_SPECIFIED_TESTS : NONE; 
```
It will be ok for checkstyle?

Tibor17 commented:
@imonteroperez
We already discussed this feature request with @olamy in ASF.
This feature bring chaos to the meaning of the config parameter and a duplicate to inclusion filter which have the space for the method filtering. Naturally, `includesFile` cannot be about filtering methods.
Instead of those parameters pls use `<test>*Test#test*One+testTwo?????</test>`.
Why you do not use the config parameter `test`? Do you want to override or merge this parameter with system property `test`?

TestListResolver

imonteroperez commented:
> @imonteroperez We already discussed this feature request with @olamy in ASF. This feature brings chaos to the meaning of the config parameter and a duplicate to filter `test` which have the space for the method filtering. Naturally, `includesFile` cannot be about filtering methods. Instead of those parameters pls use `<test>*Test#test*One+testTwo?????</test>`. Why you do not use the config parameter `test`? Do you want to override or merge this parameter with system property `test`?

Hi @Tibor17 I know I can use `-Dtest`, but I was looking for a solution to expand the method filtering capabilities to be used on `surefire.excludesFile` and (additionally) on `surefire.includesFile`.

imonteroperez commented:
Moving this PR to draft status to achieve some of the feedback provided and to include more testing.

olamy commented:
> @imonteroperez We already discussed this feature request with @olamy in ASF. This feature brings chaos to the meaning of the config parameter and a duplicate to filter `test` which have the space for the method filtering. Naturally, `includesFile` cannot be about filtering methods. Instead of those parameters pls use `<test>*Test#test*One+testTwo?????
</test>`. Why you do not use the config parameter `test`? Do you want to override or merge this parameter with system property `test`?


`Naturally`? the parameter is called `includesFile` and not `includesClassesFile` so naturally it would make sense to include methods filtering as well. the parameter `test` does `naturally`?
At the moment it's not consistent not being able to have method filtering when using files.
Naturally it's a good idea to have same level of filtering support for both parameter and it will be more consistent.
very long command line parameter are not easy to use. 
@Tibor17 I don't see any problem make parameters consistent.

Tibor17 commented:
@imonteroperez
@slawekjaranowski
@olamy

The includes|excludes, includesFile|excludesFile end up within the TestListResolver.
There is no need to modify the implementation of Provider, TestListResolver and other stuff.
All we could do is to compromise the sanity check `Method filter prohibited in includes|excludes|includesFile|excludesFile parameter`.

Below you can see the original method for includesFile. And then the sanity check is shifted in the modified code, this is my proposal.

```
    private List<String> getIncludeList()
        throws MojoFailureException
    {
        List<String> includes = null;
        if ( isSpecificTestSpecified() )
        {
            includes = new ArrayList<>();
            addAll( includes, split( getTest(), "," ) );
        }
        else
        {
            if ( getIncludesFile() != null )
            {
                includes = readListFromFile( getIncludesFile() );
            }

            if ( includes == null )
            {
                includes = getIncludes();
            }
            else
            {
                maybeAppendList( includes, getIncludes() );
            }

            checkMethodFilterInIncludesExcludes( includes );

            if ( includes == null || includes.isEmpty() )
            {
                includes = asList( getDefaultIncludes() );
            }
        }

        return filterNulls( includes );
    }
```

changed to

```
    private List<String> getIncludeList()
        throws MojoFailureException
    {
        List<String> includes = null;
        if ( isSpecificTestSpecified() )
        {
            includes = new ArrayList<>();
            addAll( includes, split( getTest(), "," ) );
        }
        else
        {
            includes = getIncludes();

            checkMethodFilterInIncludesExcludes( includes );

            if ( getIncludesFile() != null )
            {
                includes = readListFromFile( getIncludesFile() );
            }

            if ( includes == null || includes.isEmpty() )
            {
                includes = asList( getDefaultIncludes() );
            }
        }

        return filterNulls( includes );
    }
```

Similar with the excludesFile.

@imonteroperez
Would this help you?
Of course we do not know the IT results yet. They may fail.., not sure!
Of course the JavaDoc of both parameters should say that the pattern allows using #method, and the site documentation too.

imonteroperez commented:
> @imonteroperez @slawekjaranowski @olamy
> 
> The includes|excludes, includesFile|excludesFile end up within the TestListResolver. There is no need to modify the implementation of Provider, TestListResolver and other stuff. All we could do is to compromise the sanity check `Method filter prohibited in includes|excludes|includesFile|excludesFile parameter`.
> 
> Below you can see the original method for includesFile. And then the sanity check is shifted in the modified code, this is my proposal.
> 
> ```
>     private List<String> getIncludeList()
>         throws MojoFailureException
>     {
>         List<String> includes = null;
>         if ( isSpecificTestSpecified() )
>         {
>             includes = new ArrayList<>();
>             addAll( includes, split( getTest(), "," ) );
>         }
>         else
>         {
>             if ( getIncludesFile() != null )
>             {
>                 includes = readListFromFile( getIncludesFile() );
>             }
> 
>             if ( includes == null )
>             {
>                 includes = getIncludes();
>             }
>             else
>             {
>                 maybeAppendList( includes, getIncludes() );
>             }
> 
>             checkMethodFilterInIncludesExcludes( includes );
> 
>             if ( includes == null || includes.isEmpty() )
>             {
>                 includes = asList( getDefaultIncludes() );
>             }
>         }
> 
>         return filterNulls( includes );
>     }
> ```
> 
> changed to
> 
> ```
>     private List<String> getIncludeList()
>         throws MojoFailureException
>     {
>         List<String> includes = null;
>         if ( isSpecificTestSpecified() )
>         {
>             includes = new ArrayList<>();
>             addAll( includes, split( getTest(), "," ) );
>         }
>         else
>         {
>             includes = getIncludes();
> 
>             checkMethodFilterInIncludesExcludes( includes );
> 
>             if ( getIncludesFile() != null )
>             {
>                 if ( includes == null )
>                 {
>                     includes = new ArrayList<>();
>                 }
>                 addAll( includes, readListFromFile( getIncludesFile() ) );
>             }
> 
>             if ( includes == null || includes.isEmpty() )
>             {
>                 includes = asList( getDefaultIncludes() );
>             }
>         }
> 
>         return filterNulls( includes );
>     }
> ```
> 
> Similar with the excludesFile.
> 
> @imonteroperez Would this help you? Of course we do not know the IT results yet. They may fail.., not sure! The JavaDoc of both parameters should say that the pattern allows using #method, and the site documentation too.

I think this is not going to work because IIUC then is going to be delegated to a FileScanner to evaluate which test classes should be executed without considering the method filtering. See: https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/util/FileScanner.java#L74 where method is `null`

imonteroperez commented:
> > @imonteroperez @slawekjaranowski @olamy
> > The includes|excludes, includesFile|excludesFile end up within the TestListResolver. There is no need to modify the implementation of Provider, TestListResolver and other stuff. All we could do is to compromise the sanity check `Method filter prohibited in includes|excludes|includesFile|excludesFile parameter`.
> > Below you can see the original method for includesFile. And then the sanity check is shifted in the modified code, this is my proposal.
> > ```
> >     private List<String> getIncludeList()
> >         throws MojoFailureException
> >     {
> >         List<String> includes = null;
> >         if ( isSpecificTestSpecified() )
> >         {
> >             includes = new ArrayList<>();
> >             addAll( includes, split( getTest(), "," ) );
> >         }
> >         else
> >         {
> >             if ( getIncludesFile() != null )
> >             {
> >                 includes = readListFromFile( getIncludesFile() );
> >             }
> > 
> >             if ( includes == null )
> >             {
> >                 includes = getIncludes();
> >             }
> >             else
> >             {
> >                 maybeAppendList( includes, getIncludes() );
> >             }
> > 
> >             checkMethodFilterInIncludesExcludes( includes );
> > 
> >             if ( includes == null || includes.isEmpty() )
> >             {
> >                 includes = asList( getDefaultIncludes() );
> >             }
> >         }
> > 
> >         return filterNulls( includes );
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > changed to
> > ```
> >     private List<String> getIncludeList()
> >         throws MojoFailureException
> >     {
> >         List<String> includes = null;
> >         if ( isSpecificTestSpecified() )
> >         {
> >             includes = new ArrayList<>();
> >             addAll( includes, split( getTest(), "," ) );
> >         }
> >         else
> >         {
> >             includes = getIncludes();
> > 
> >             checkMethodFilterInIncludesExcludes( includes );
> > 
> >             if ( getIncludesFile() != null )
> >             {
> >                 if ( includes == null )
> >                 {
> >                     includes = new ArrayList<>();
> >                 }
> >                 addAll( includes, readListFromFile( getIncludesFile() ) );
> >             }
> > 
> >             if ( includes == null || includes.isEmpty() )
> >             {
> >                 includes = asList( getDefaultIncludes() );
> >             }
> >         }
> > 
> >         return filterNulls( includes );
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Similar with the excludesFile.
> > @imonteroperez Would this help you? Of course we do not know the IT results yet. They may fail.., not sure! The JavaDoc of both parameters should say that the pattern allows using #method, and the site documentation too.
> 
> I think this is not going to work because IIUC then is going to be delegated to a FileScanner to evaluate which test classes should be executed without considering the method filtering. See: https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/util/FileScanner.java#L74 where method is `null`

Applied your proposal with a combination of mine to specify the tests to run based on identifying if there is method filtering applied or not on the `TestListResolver`

slawekjaranowski commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java:
```diff
@@ -2292,23 +2291,52 @@ private void checkMethodFilterInIncludesExcludes( Iterable<String> patterns )
     }
 
-    private TestListResolver getIncludedAndExcludedTests()
+    private TestListResolver solveIncludedAndExcludedTests()
```
Method returns two values .... 

One by standard `return` statement and second by call `setTest(...)` which changes `Mojo` state and can cause impact to other place ... 

I didn't analyze flow where `solveIncludedAndExcludedTests` and `getTest` are used and what order are.

Try to describe what does this method do ...

imonteroperez replied:
Updated in https://github.com/apache/maven-surefire/pull/400/commits/bc8768bccda866ee3edbb7c12a8f439ff351bcdd

slawekjaranowski commented:
code looks ok 

Please add documentation to Mojo params and  to site with examples.

Tibor17 commented:
@imonteroperez 
I did not mean to put the change on this branch. Here are too many and unnecessary changes.
The fix should be quite simple. The lists are merged in those two methods we know and no other parts of the code should be modified.
Pls try to modify the methods `getIncludeList` and `getExcludeList` on the master. There is no reason why it should not work with minimum change.

Tibor17 commented:
@imonteroperez
@slawekjaranowski
Pls let a have a look into #440 as well.

imonteroperez commented:
Closed in favor of https://github.com/apache/maven-surefire/pull/440

jira-importer commented:
Resolve #2883
